crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2199,33 +2199,65 @@ impl<'db> Type<'db> {
             // TODO annotated return type on `__new__` or metaclass `__call__`
             // TODO check call vs signatures of `__new__` and/or `__init__`
             Type::ClassLiteral(ClassLiteralType { class }) => {
+                let return_ty = match class.known(db) {
+                    // TODO: We should check the call signature and error if the bool call doesn't have the
+                    //   right signature and return a binding error.
+
+                    // If the class is the builtin-bool class (for example `bool(1)`), we try to
+                    // return the specific truthiness value of the input arg, `Literal[True]` for
+                    // the example above.
+                    Some(KnownClass::Bool) => arguments
+                        .first_argument()
+                        .map(|arg| arg.bool(db).into_type(db))
+                        .unwrap_or(Type::BooleanLiteral(false)),
+
+                    // TODO: Don't ignore the second and third arguments to `str`
+                    //   https://github.com/astral-sh/ruff/pull/16161#discussion_r1958425568
+                    Some(KnownClass::Str) => arguments
+                        .first_argument()
+                        .map(|arg| arg.str(db))
+                        .unwrap_or_else(|| Type::string_literal(db, "")),
+
+                    Some(KnownClass::Type) => arguments
+                        .exactly_one_argument()
+                        .map(|arg| arg.to_meta_type(db))
+                        .unwrap_or_else(|| KnownClass::Type.to_instance(db)),
+
+                    // TODO: check signatures of other KnownClass constructors
+                    Some(_) => Type::Instance(InstanceType { class }),
```

Comment:
Could this `Some(_)` path be merged with the `None` path below, i.e. couldn't we rely on typeshed stubs to give us the right `__init__` signature for these classes? Do we need the `class.known`-specialization at all?

Reply:
> doesn't look like so. E.g. the tests that involved `int` failed without this branch, which means typeshed lacks the proper signature or something else going on.

You're right, I guess we can only do that once we also check the `__new__` calls (and understand overloads)

https://github.com/astral-sh/ruff/blob/b3c884f4f33a81cd08e9ebc13fd0d5a538c8afc7/crates/red_knot_vendored/vendor/typeshed/stdlib/builtins.pyi#L235-L239

> I intended to close this PR without touching known classes


Sounds good to me.